go-runtime/ftl/typeregistry/type_registry.go

```diff
@@ -13,46 +14,88 @@ func ContextWithTypeRegistry(ctx context.Context, r *TypeRegistry) context.Conte
 	return context.WithValue(ctx, contextKeyTypeRegistry{}, r)
 }
 
-// TypeRegistry is a registry of types that can be instantiated by their qualified name.
-// It also records sum types and their variants, for use in encoding and decoding.
+// FromContext retrieves the secrets schema.TypeRegistry previously
+// added to the context with [ContextWithTypeRegistry].
+func FromContext(ctx context.Context) optional.Option[*TypeRegistry] {
+	t, ok := ctx.Value(contextKeyTypeRegistry{}).(*TypeRegistry)
+	if ok {
+		return optional.Some(t)
+	}
+	return optional.None[*TypeRegistry]()
+}
+
+// TypeRegistry is used for dynamic type resolution at runtime. It stores associations between sum type discriminators
+// and their variants, for use in encoding and decoding.
 //
-// FTL manages the type registry for you, so you don't need to create one yourself
+// FTL manages the type registry for you, so you don't need to create one yourself.
 type TypeRegistry struct {
-	// GoTypes associates a type name with a Go type.
-	GoTypes map[string]reflect.Type
-	// SumTypes associates a sum type discriminator type name with its variant type names.
-	SumTypes map[string][]string
+	sumTypes map[string][]sumTypeVariant
+}
+
+type sumTypeVariant struct {
+	name   string
+	goType reflect.Type
 }
 
 // NewTypeRegistry creates a new type registry.
 // The type registry is used to instantiate types by their qualified name at runtime.
 func NewTypeRegistry() *TypeRegistry {
 	return &TypeRegistry{
-		GoTypes:  make(map[string]reflect.Type),
-		SumTypes: make(map[string][]string),
-	}
-}
-
-// New creates a new instance of the type from the qualified type name.
-func (t *TypeRegistry) New(name string) (any, error) {
-	typ, ok := t.GoTypes[name]
-	if !ok {
-		return nil, fmt.Errorf("type %q not registered", name)
+		sumTypes: make(map[string][]sumTypeVariant),
 	}
-	return reflect.New(typ).Interface(), nil
 }
 
 // RegisterSumType registers a Go sum type with the type registry. Sum types are represented as enums in the
 // FTL schema.
 func (t *TypeRegistry) RegisterSumType(discriminator reflect.Type, variants map[string]reflect.Type) {
```

Comment:
What are they keys to `variants`? The unqualified FTL type names, Go type names?

Reply:
the keys are the variants names, e.g.:
```
typeRegistry.RegisterSumType(reflect.TypeFor[foo.TypeEnum](), map[string]reflect.Type{
		"A": reflect.TypeFor[foo.A](),
		"B": reflect.TypeFor[foo.B](),
})
```

fwiw this part is only used by generated code!